File: 45-jump-game-ii/45-jump-game-ii.java
class Solution {
    Integer[] memo;
    
    public int jump(int[] nums) {
        int n = nums.length;
        memo = new Integer[n];
        return helper(nums, 0);
    }
    public int helper(int[] nums, int start) {
        //base case
        if(start >= nums.length - 1) return 0;
        if(memo[start] != null) return memo[start];
        int step = nums.length;//注意step的值
        int next = Math.min(nums.length - 1, nums[start] + start);
        for(int i = start + 1; i <= next; i++) {
            step = Math.min(step, 1 + helper(nums, i));
        }
        return memo[start] = step;
    }
    
}

File: 45-jump-game-ii/45-jump-game-ii.py
class Solution:
    def jump(self, nums: List[int]) -> int:
        dp = [0] * len(nums)
        j = 0
        for i in range(1, len(nums)):
            while j + nums[j] < i:
                j += 1
            dp[i] = dp[j] + 1
        return dp[len(nums) - 1]